libgnucash/engine/mocks/fake-qofquery.cpp

```diff
@@ -1,46 +1,112 @@
 #include <config.h>
 
+#include <list>
+
 #include "fake-qofquery.h"
 #include "gmock-qofbook.h"
 
+
+
+/* class QofFakeQueryPool */
+
+static class QofFakeQueryPool
+{
+public:
+    QofFakeQueryPool() {};
```

Comment:
Ok, then I would omit the definition. This has to be changed for the constructor of class `MockAccount` as well, right?

Reply:
Yes.

Reply:
That aspect was very interesting. I removed the empty constructor for classes `QofFakeQueryPool` and `MockPrefsBackend` and everything was fine.

But after removing the empty constructor for class `MockAccount`, the test application crashed with a `SIGSEV` in the `delete` operator. What happened?

I debugged this and found out, that the default constructor does zero initialization and the empty constructor does not. And since the constructor runs *after* the `new` operator, which does `gobject` initialization, the default constructor overwrites all `gobject` data, but the empty constructor does not.

cppreference says about [value initialization](https://en.cppreference.com/w/cpp/language/value_initialization) since C++11: *"if T is a class type with no default constructor or with a user-provided or deleted default constructor, the object is default-initialized"*, but *"if T is a class type with a default constructor that is neither user-provided nor deleted (that is, it may be a class with an implicitly-defined or defaulted default constructor), the object is zero-initialized and then it is default-initialized if it has a non-trivial default constructor"*

So we can not remove the empty constructor here.

Reply:
Interesting. That either needs a comment or a second think about doing GObject initialization in operator new (and I suppose dispose/finalize in operator delete). As I'm sure you know, operators new and delete are for allocating and freeing memory and initialization/finalization are supposed to happen in the constructor and destructor.

Reply:
Yes, I know that. But since `g_object_new()` does allocation and initialization in one, you can not simply separate both, except of reimplementing parts of this function, which is challenging after looking into glib code.

On the other hand at least the correct order of allocation and initialization is fulfilled. At first memory is allocated. And the first step of object construction would be the initialization of the base class anyway. Since `GObject` can be seen as the "base class", everything is fine. The next step is the construction of the derived class, which is done by the C++ ctor. The only problem is, that this base class is hidden to the C++ compiler. Therefore some coding rules have to be followed as for instance using empty constructors instead of default constructors.

I don't see another opportunity at the moment except of reimplementing `g_object_new()` (and `g_object_unref()`).

I'll add a comment.